openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,352 @@
+ * Some globally cached values.
+ */
+static jclass OCSP_CALLBACK_CLAZZ = NULL;
+static jmethodID OCSP_CALLBACK_METHOD = NULL;
+
+/**
+ * Frees the callback
+ */
+static void openssl_netty_tcnative_ocsp_free_callback(JNIEnv *env, tcn_ssl_ctxt_t *c, jweak callbackRef)
+{
+    // Free the global weak reference of the callback
+    (*env)->DeleteWeakGlobalRef(env, callbackRef);
+
+    if (c != NULL) {
+        SSL_CTX_set_tlsext_status_arg(c->ctx, NULL);
+        SSL_CTX_set_tlsext_status_cb(c->ctx, NULL);
+    }
+}
+
+/**
+ * This function is being called from OpenSSL.
+ *
+ * See:
+ *     openssl_netty_tcnative_ocsp_client_callback
+ *     openssl_netty_tcnative_ocsp_server_callback
+ */
+static int openssl_netty_tcnative_ocsp_callback(int error, SSL *ssl, void *arg)
+{
+    JNIEnv *env;
+    if (tcn_get_java_env(&env) != JNI_OK) {
+        fprintf(stderr, "openssl_netty_tcnative_ocsp_callback: no JNIEnv\n");
+        return error;
+    }
+
+    if (OCSP_CALLBACK_CLAZZ == NULL) {
+        tcn_ThrowException(env, "openssl_netty_tcnative_ocsp_callback: no callback class");
+        return error;
+    }
+
+    if (OCSP_CALLBACK_METHOD == NULL) {
+        tcn_ThrowException(env, "openssl_netty_tcnative_ocsp_callback: no callback method");
+        return error;
+    }
+
+    if (arg == NULL) {
+        tcn_ThrowException(env, "openssl_netty_tcnative_ocsp_callback: no callback args");
+        return error;
+    }
+
+    int response = 0;
+    jobject callback = (*env)->NewLocalRef(env, (jweak)arg);
+
+    if ((*env)->IsSameObject(env, callback, NULL) == JNI_FALSE) {
+        jlong ssl_ = P2J(ssl);
+        response = (int)(*env)->CallIntMethod(env, callback, OCSP_CALLBACK_METHOD, ssl_);
+
+        if ((*env)->ExceptionCheck(env)) {
+            // Use the given error code in the event of a JVM exception
+            response = error;
+        }
+    } else {
+        response = error;
+    }
+
+    (*env)->DeleteLocalRef(env, callback);
+    return response;
+}
+
+/**
+ * Client and Server callbacks use different error codes.
+ *
+ * @see openssl_netty_tcnative_ocsp_callback
+ */
+static int openssl_netty_tcnative_ocsp_client_callback(SSL *ssl, void *arg)
+{
+    return openssl_netty_tcnative_ocsp_callback(TCN_CLIENT_CALLBACK_ERROR, ssl, arg);
+}
+
+/**
+ * Client and Server callbacks use different error codes.
+ *
+ * @see openssl_netty_tcnative_ocsp_callback
+ */
+static int openssl_netty_tcnative_ocsp_server_callback(SSL *ssl, void *arg)
+{
+    return openssl_netty_tcnative_ocsp_callback(TCN_SERVER_CALLBACK_ERROR, ssl, arg);
+}
+#endif
+
+/**
+ * Returns true is OCSP stapling is available.
+ *
+ * NOTE: There can be basically three reasons why OCSP stapling
+ * may not be available:
+ *
+ * 1) It's outright not supported (e.g. OpenSSL < 1.0.0).
+ * 2) OpenSSL was explicitly built without OCSP stapling support.
+ * 3) It's supported but not implemented by us (e.g. BoringSSL)
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, isOcspAvailable)(TCN_STDARGS)
```

Comment:
On further thought and looking at OpenSSL's code (e.g. [tls1.h](https://github.com/openssl/openssl/blob/master/include/openssl/tls1.h#L282) / `SSL_set_tlsext_status_type` and the impl. in [s3_lib.c](https://github.com/openssl/openssl/blob/master/ssl/s3_lib.c#L3097)) and how the `OPENSSL_NO_OCSP` ifdef is being used... I think I'm overthinking it here.

These [functions](https://www.openssl.org/docs/man1.0.0/apps/ocsp.html) [\[alt-link\]](http://manpages.ubuntu.com/manpages/xenial/man3/SSL_CTX_set_tlsext_status_cb.3ssl.html) are always present as of OpenSSL >= 1.0.0. 

Bottom line, we don't need this extra function and we can simply check for OpenSSL >= 1.0.0 like the ALPN thing does.